frame/state-trie-migration/src/lib.rs

```diff
@@ -0,0 +1,1417 @@
+				},
+				(Some(ref top_key), None) => {
+					// we have a top key and no child key. 3 possibilities exist:
+					// 1. we continue the top key migrations.
+					// 2. this is the root of a child key, and we start processing child keys (and
+					// should call `migrate_child`).
+					// 3. this is the root of a child key, and we are finishing all child-keys (and
+					// should call `migrate_top`).
+
+					// NOTE: this block is written intentionally to verbosely for easy of
+					// verification.
+					match (
+						top_key.starts_with(DEFAULT_CHILD_STORAGE_KEY_PREFIX),
+						self.prev_tick_child,
+					) {
+						(false, false) => {
+							// continue the top key migration
+							self.migrate_top();
+						},
+						(true, false) => {
+							// start going into a child key. In the first iteration, we always
+							let maybe_first_child_key = {
+								// just in case there's some data in `&[]`, read it. Since we can't
+								// check this without reading the actual key, and given that this
+								// function should always read at most one key, we return after
+								// this. The rest of the migration should happen in the next tick.
+								let child_top_key = Pallet::<T>::child_io_key_or_halt(top_key);
+								let _ = sp_io::default_child_storage::get(child_top_key, &vec![]);
+								sp_io::default_child_storage::next_key(child_top_key, &vec![])
+							};
+							if let Some(first_child_key) = maybe_first_child_key {
+								self.current_child = Some(first_child_key);
+								self.prev_tick_child = true;
+							} else {
+								// we have already done a (pretty useless) child key migration, just
+								// set the flag. Since we don't set the `self.current_child`, next
+								// tick will move forward to the next top key.
+								self.prev_tick_child = true;
+							}
+						},
+						(true, true) => {
+							// we're done with migrating a child-root.
+							self.prev_tick_child = false;
+							self.migrate_top();
+						},
+						(false, true) => {
+							// should never happen.
+							log!(error, "LOGIC ERROR: unreachable code [0].");
+							Pallet::<T>::halt();
+						},
+					};
+				},
+				(None, Some(_)) => {
+					log!(error, "LOGIC ERROR: unreachable code [1].");
+					Pallet::<T>::halt()
+				},
+				(None, None) => {
+					// nada
+				},
+			}
+		}
+
+		/// Migrate the current child key, setting it to its new value, if one exists.
+		///
+		/// It updates the dynamic counters.
+		fn migrate_child(&mut self) {
+			let last_child = self.current_child.as_ref().expect("value checked to be `Some`; qed");
+			let last_top = self.current_top.clone().expect("value checked to be `Some`; qed");
+
+			let child_root = Pallet::<T>::child_io_key_or_halt(&last_top);
+			let added_size =
+				if let Some(data) = sp_io::default_child_storage::get(child_root, &last_child) {
+					self.dyn_size = self.dyn_size.saturating_add(data.len() as u32);
+					sp_io::default_child_storage::set(child_root, last_child, &data);
```

Comment:
Why do you need to write it again? I thought reading already migrates.

Reply:
No, afaik to migrate each key you need to read and write it.